OutOfSchool/OutOfSchool.WebApi/Services/AverageRatings/AverageRatingService.cs

```diff
@@ -0,0 +1,229 @@
+
+        logger.LogInformation("Getting the average rating by workshop's or provider's id finished.");
+
+        return mapper.Map<AverageRatingDto>(rating);
+    }
+
+    /// <inheritdoc/>
+    public async Task<IEnumerable<AverageRatingDto>> GetByEntityIdsAsync(IEnumerable<Guid> entityIds)
+    {
+        logger.LogInformation("Getting the average ratings by the list of the workshop's or provider's ids started.");
+
+        var ratings = await averageRatingRepository.GetByFilter(r => entityIds.Contains(r.EntityId)).ConfigureAwait(false);
+
+        logger.LogInformation("Getting the average ratings by the list of the workshop's or provider's ids finished.");
+
+        return mapper.Map<IEnumerable<AverageRatingDto>>(ratings);
+    }
+
+    /// <inheritdoc/>
+    public async Task CalculateAsync(CancellationToken cancellationToken = default)
+    {
+        logger.LogInformation("The average ratings calculation was started.");
+
+        var lastSuccessRatingsCalculationDate = await GetLastSuccessRatingCalculationDateAsync().ConfigureAwait(false);
+        var currentRatingsCalculationDate = DateTimeOffset.UtcNow;
+
+        var newAddedRatings = await ratingService.GetAllAsync(r => r.CreationTime > lastSuccessRatingsCalculationDate).ConfigureAwait(false);
+
+        if (newAddedRatings.Any())
+        {
+            var averageRatings = await RecalculateAverageRatingsAsync(newAddedRatings).ConfigureAwait(false);
+
+            var executionStrategy = db.Database.CreateExecutionStrategy();
+            await executionStrategy.Execute(SaveAverageRatings).ConfigureAwait(false);
+
+            async Task SaveAverageRatings()
+            {
+                await using IDbContextTransaction transaction = await db.Database.BeginTransactionAsync(cancellationToken).ConfigureAwait(false);
+
+                try
+                {
+                    await SaveAverageRatinsgAsync(averageRatings, currentRatingsCalculationDate).ConfigureAwait(false);
+                    await transaction.CommitAsync(cancellationToken);
+                }
+                catch (Exception ex)
+                {
+                    await transaction.RollbackAsync(cancellationToken);
+                    logger.LogError("The average ratings calculation error. The average ratings weren't saved.");
+                    throw;
+                }
+            }
+        }
+
+        logger.LogInformation("The average ratings calculation was finished.");
+    }
+
+    private async Task<DateTimeOffset> GetLastSuccessRatingCalculationDateAsync()
+    {
+        logger.LogInformation("Getting the date of the last success rating calculation started.");
+
+        var ratings = await averageRatingRepository.GetAll().ConfigureAwait(false);
+
+        logger.LogInformation("Getting the date of the last success rating calculation finished.");
+
+        return ratings.Any() ? ratings.Max(r => r.CreationTime) : DateTimeOffset.MinValue;
+    }
```

Comment:
Is it faster than one-row table for storing job's data? We get all rows and calculate max value after that.

Reply:
I've done it, and made an additional table QuartzJobs